src/utilities/geometry/Polyhedron.cpp

```diff
@@ -58,136 +67,246 @@ bool Surface3dEdge::operator!=(const Surface3dEdge& other) const {
 }
 
+Vector3d Surface3dEdge::asVector() const {
+  return m_end - m_start;
+}
+
 std::ostream& operator<<(std::ostream& os, const Surface3dEdge& edge) {
   os << "Surface3dEdge: start=" << edge.start() << ", end=" << edge.end() << ", count=" << edge.count()
-     << ", firstSurface=" << edge.allSurfaces()[0].name;
+     << ", firstSurface=" << edge.firstSurfaceName();
   return os;
 }
 
-Surface3d::Surface3d(std::vector<Point3d> t_vertices, std::string t_name) : vertices(std::move(t_vertices)), name(std::move(t_name)) {}
+Surface3d::Surface3d(std::vector<Point3d> t_vertices, std::string t_name, size_t t_surfNum)
+  : vertices(std::move(t_vertices)), name(std::move(t_name)), surfNum(t_surfNum) {
+  for (auto it = vertices.begin(); it != vertices.end(); ++it) {
 
-Polyhedron::Polyhedron(std::vector<Surface3d> surfaces) : m_surfaces(std::move(surfaces)){};
+    auto itnext = std::next(it);
+    if (itnext == std::end(vertices)) {
+      itnext = std::begin(vertices);
+    }
 
-size_t Polyhedron::numVertices() const {
-  size_t count = 0;
-  for (const auto& surface : m_surfaces) {
-    count += surface.vertices.size();
+    edges.emplace_back(*it, *itnext, t_name, t_surfNum);
   }
-  return count;
 }
 
-std::vector<Point3d> Polyhedron::uniqueVertices() const {
+bool Surface3d::operator<(const Surface3d& rhs) const {
+  return this->name < rhs.name;
+}
 
-  std::vector<Point3d> uniqVertices;
-  uniqVertices.reserve(numVertices());
+size_t Surface3d::numConflictedEdges() const {
+  return std::count_if(edges.cbegin(), edges.cend(), [](const auto& edge) { return edge.hasConflictedOrientation(); });
+}
 
-  for (const auto& surface : m_surfaces) {
-    for (const auto& pt : surface.vertices) {
-      if (std::find_if(uniqVertices.cbegin(), uniqVertices.cend(), [&pt](const auto& unqV) { return isAlmostEqual3dPt(pt, unqV); })
-          == uniqVertices.cend()) {
-        uniqVertices.push_back(pt);
+double Surface3d::ratioOfConflictedEdges() const {
+  return static_cast<double>(numConflictedEdges()) / static_cast<double>(edges.size());
+}
+
+void Surface3d::resetEdgeMatching() {
+  for (auto& edge : edges) {
+    edge.resetEdgeMatching();
+  }
+}
+
+bool Surface3d::isConvex() const {
+
+  // const auto& a = edges.front().start();
+  // const auto& b = edges.front().end();
+  // const auto& c = edges.at(1).end();
+
+  auto ab = edges.front().asVector();
+  auto bc = edges[1].asVector();
+  auto outwardNormal = ab.cross(bc);
+  outwardNormal.normalize();
+
+  for (auto it = std::next(edges.begin()); it != edges.end(); ++it) {
+    auto itnext = std::next(it);
+    if (itnext == std::end(edges)) {
+      itnext = std::begin(edges);
+    }
+    const Point3d& a = it->start();
+    const Point3d& b = it->end();
+    const Point3d& c = itnext->end();
+
+    auto ab = b - a;
+    auto ac = c - a;
+
+    auto triangleNormal = ab.cross(ac);
+    auto d = outwardNormal.dot(triangleNormal);
+    if (d < 0) {
+      return false;
+    }
+  }
+  return true;
+}
```

Comment:
Convexity of a Surface3d: for each triangle of 3 points, we compute the triangleNormal by taking the cross product of AB x AC. We then take the dot product with the initial outwardNormal. If the dot product is negative, this means the triangleNomal is pointing in the opposite direction and we therefore have a non convex / concave surface.